src/huggingface_hub/inference/_providers/_common.py

```diff
@@ -135,15 +136,17 @@ def _prepare_headers(self, headers: Dict, api_key: str) -> Dict:
 
         Override this method in subclasses for customized headers.
         """
-        return {**build_hf_headers(token=api_key), **headers}
+        self.headers = {**build_hf_headers(token=api_key), **headers}
+        return self.headers
 
     def _prepare_url(self, api_key: str, mapped_model: str) -> str:
         """Return the URL to use for the request.
 
         Usually not overwritten in subclasses."""
         base_url = self._prepare_base_url(api_key)
         route = self._prepare_route(mapped_model)
-        return f"{base_url.rstrip('/')}/{route.lstrip('/')}"
+        self.url = f"{base_url.rstrip('/')}/{route.lstrip('/')}"
```

Comment:
I'm worried about that :confused: 
The client could be used with multithreading, making multiple calls in parallel to multiple models. Saving `self.url` as class instance is dangerous. Same for `self.api_key` :grimacing: (see below). It should be fine in 99% but one day we might end up with a very annoying bug to investigate.